### ISSUE TITLE
[slurm] wandb hangs at the end of jobs in dryrun mode

### ISSUE THREAD

**Issue by bknyaz** [wandb/wandb#919]
`wandb --version && python --version && uname`

* Weights and Biases version:  0.8.21
* Python version: Python 3.6.8 :: Anaconda, Inc.
* Operating System: CentOS Linux release 7.7.1908 (Core)

### Description

I'm using wandb on the GPU cluster with slurm to run jobs. 
After the script finishes, wandb prints the following:

```
wandb: Waiting for W&B process to finish, PID {some process id}
wandb: Program ended successfully.
```

The problem is that the slurm scheduler doesn't quit this job and occupies the GPU node. Perhaps, for some reason some wandb processes are still running?

Not sure if the issue is with wandb or with the cluster I'm using. The cluster is actually one of the biggest in Canada, so I can imagine other people have this issue and it can result in a lot of nodes being idle for no reason. So would be great to solve this.

Other clusters I've used with Ubuntu and Internet access worked fine.

I use WANDB_MODE=dryrun, because the cluster doesn't have access to external network. 

**Update**
My impression is that wandb tries to connect to the server after the script is finished, but because 
there is no connection, it raises some exception and the process gets stuck for some reason.

In one of my log files I found an additional line printed at the end regarding the connection:
```
wandb: Waiting for W&B process to finish, PID {some process id}
wandb: Program ended successfully.
wandb: ERROR Failed to connect to W&B. Retrying in the background.
```

### What I Did

see above

Thanks.

**Comment by cvphelps** [comment]
Thanks for reporting this! @raubitsj could you please take a look

**Comment by ariG23498** [comment]
Hey @bknyaz 
In the past year we've majorly reworked the CLI and UI for Weights & Biases. We're closing issues older than 6 months. Please comment to reopen.

**Comment by vildera** [comment]
Did anyone find a solution to this? I have the same problem.

**Comment by lukekenworthy** [comment]
I am having this problem as well. Has anyone ever figured this out?

**Comment by vanpelt** [comment]
@lukekenworthy can you provide an example script?  If you're using multiprocessing in your scripts you may need to explicity call `wandb.finish()` in the process that called wandb.init once processing as completed.

**Comment by mohamedr002** [comment]
> @lukekenworthy can you provide an example script? If you're using multiprocessing in your scripts you may need to explicitly call `wandb.finish()` in the process that called wandb.init once processing as completed.

I am having same issue when using wandb.sweep, where can I put wandb.finish() exactly in the script?

**Comment by ssadhukha** [comment]
I'm dealing with the same issue, any solutions for this yet?

**Comment by zdhNarsil** [comment]
Same issue here. Seems no update?

**Comment by vanpelt** [comment]
@ssadhukha or @zdhNarsil can you share an example script that gets you into this state?

**Comment by davidoort** [comment]
Same issue here

**Comment by ssadhukha** [comment]
I'm running into this state when I specify a different run inside a for loop. The main parts of my script below:


```
## IMPORT PACKAGES + LOCAL MODULES

import os
import sys
import torch
from torch.utils.data import DataLoader
from torch.utils.data import Subset
import pytorch_lightning as pl  # lightning toolbox
from pytorch_lightning.loggers import WandbLogger
import wandb

wandb.login()
wandb_logger = WandbLogger(project="transformer-tcg", log_model=True)
pl.seed_everything(42)

sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), "../..")))
from tools.transformer_tcg import *
from tools.analysis_utils import *
from tools.dataset import *

## FUNCTIONS

def main():
    run = wandb.init(name=f"pair{int(pr)}-{dtype}", project="transformer", allow_val_change=True)
    print(f"Num of layers: {nlayers}")
    print(f"Num of heads: {nheads}")
    basename = (
        f"pr{int(pr)}_trl_{trl}_{dtype}_{nlayers}layers_{nheads}heads_{ndims}dims_"
    )
    models = glob(os.path.join(pr_modeldir, "%s*" % basename + "*weights"))
    if models == []:
        ## For the first trained model
        filename = basename + "0000"
    else:
        last_num = max([int(i.split("_")[-2]) for i in models])
        filename = basename + f"{str(last_num + 1).zfill(4)}"

    filename = os.path.join(pr_modeldir, filename)

    if not os.path.isfile(filename + "_weights"):
        print("Training " + filename)
        model, trainer = train(
            train_loader,
            val_loader,
            filename,
            pretrain_filename=pretrain_filename,
            early_stopping_method=early_stopping_method,
            patience=patience,
            verbose=None,
            max_epochs=max_epochs,
            mode=mode,
            min_delta=min_delta,
            input_dim=nfeats,
            max_len=maxlen,
            num_classes=nclasses,
            model_dim=ndims,
            num_heads=nheads,
            num_layers=nlayers,
            max_pool=False,
            dropout=dropout,
            lr=lr,
            warmup=warmup,
            min_acc=min_acc,
            cross_val=cross_val,
            logger=False,
            stop_early=stop_early,
        )
        run.finish()
    return model, trainer, filename

## MAIN SCRIPT
# --------------------------------- Model parameters --------------------------------- #
###
###
### ...

dtype="test"
model_type="test"
projectdir="test"
pairs = [i for i in range(1,27)]
trials = [i for i in range(1,80)]
nlayers=2
nheads=2
ndims=16
for pr in pairs:
    print(f"Pair: {pr}")
    pr_modeldir = os.path.join(
        projectdir, "models", dtype, model_type, str(int(pr))
    )  # Save models here
    if os.path.exists(pr_modeldir):
        pass
    else:
        os.mkdir(pr_modeldir)
    for trl in trials:
        if trl == torch.tensor(1):
            pretrain_filename = None
        else:
            pretrain_filename = os.path.join(
                pr_modeldir,
                f"pr{int(pr)}_trl_{trl-1}_{dtype}_{nlayers}layers_{nheads}heads_{ndims}dims_0000",
            )
        mask = (torch.tensor(dataset.pairs) == pr) & (
            torch.tensor(dataset.trials) == trl
        )
        indices = torch.tensor(mask)
        indices = indices.nonzero().reshape(-1)
        train_set = Subset(dataset, indices)
        train_loader = DataLoader(
            train_set, batch_size=batch_size, shuffle=True, num_workers=0
        )
        if __name__ == "__main__":
            model, trainer, filename = main()


```

**Comment by alopez1999** [comment]
Same issue here, still no update?

**Comment by cherry-nancy** [comment]
same issue here, any update?

**Comment by kptkin** [comment]
@adrialopezescoriza @cherry-nancy @davidoort could you please provide a small repro so we could debug the issue and hopefully resolve it for you?

**Comment by kptkin** [comment]
@ssadhukha just to verify something in your repro: https://github.com/wandb/wandb/issues/919#issuecomment-1454844483
if on this line: `wandb_logger = WandbLogger(project="transformer-tcg", log_model=True)` you remove the `log_model` are you still seeing an issue?
Thanks!

**Comment by tristan-spaulding2** [comment]
Based on the fix in Anmol's note above, closing this issue as resolved.  Please let us know if you're still seeing it, thanks!
